src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +25,74 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result = Unit>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
  public:
   virtual ~IRTraversingVisitor() {}
 
   // Invoked before all the children of `module` is visited.
-  virtual void PreVisit(const Module& module) {}
+  virtual Result PreVisit(const Module& module) {
+    return Result();
+  }
   // Invoked after all the children of `module` is visited.
-  virtual void PostVisit(const Module& module) {}
+  virtual Result PostVisit(const Module& module, Result result) {
+    return result;
+  }
+  virtual Result FoldResult(const Module& module, Result child_result,
+                            Result result) {
```

Comment:
Documentation?  Is this for folding the children of a `module`? Or, is `module` the child?

Also, is `result` the accumulated results so far? Can you name it appropriately? Also, a different order of argument might be better? e.g., 

```suggestion
  virtual Result FoldResult(Result accumulated_result, const Module& child, Result child_result) {
```

Whatever you decide here, please apply to other methods as well.

Reply:
Thanks Gogul, this makes sense to me. I added comments and reordered the arg. Didn't change `PostVisit` as I thought it would be confusing in the `Unit` case.